### PR TITLE
Add state manager

### LINKED ISSUE #44: Add minimal state manager
depends on https://github.com/cryptoeconomicslab/plasma-rust-framework/issues/34 and https://github.com/cryptoeconomicslab/plasma-rust-framework/issues/35

## Description

See also https://github.com/plasma-group/pigi/blob/01ab31380c8058709574647ee3037d28702e4e18/packages/core/src/app/pg-core/state/state-manager.ts.

State manager applies a transaction to the whole state.
Use range store and predicate verifier.

### PR DESCRIPTION
**Pull request by syuhei176**
Add minimal StateManager implementation and spec.

Close #44 

- [x] Add state Db
- [x] Move State Manager and State Db to appropriate place

### CODE REVIEW

File: core/src/state_manager.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+use crate::data_structure::error::Error;
+use crate::data_structure::state_update::StateUpdate;
+use crate::data_structure::transaction::Transaction;
+use plasma_db::impls::rangestore::memory::RangeDbMemoryImpl;
+use plasma_db::range::Range;
+use plasma_db::traits::db::DatabaseTrait;
+use plasma_db::traits::rangestore::RangeStore;
+
+struct MockPredicatePlugin {}
+impl MockPredicatePlugin {
```
Comment (syuhei176) [inline core/src/state_manager.rs]: Now using mock "Predicate Plugin".

File: core/src/state/state_manager.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            .iter()
+            .map(|verified_state_update| {
+                MockPredicatePlugin::execute_state_transition(
+                    verified_state_update.get_state_update(),
+                    transaction,
+                )
+            })
+            .collect();
+        // new_state_updates should has same state_update
+        let new_state_update: &StateUpdate = &new_state_updates[0];
```
Comment (syuhei176) [inline core/src/state/state_manager.rs]: I'm not sure how to merge multiple `stateUpdates`. It's enough to check all outputs are the same `stateUpdate`?

File: core/src/state/state_manager.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        let transaction = Transaction::new(
+            Address::zero(),
+            0,
+            100,
+            Transaction::create_method_id(&b"send(address)"[..]),
+            &b"new state update"[..],
+            &Witness::new(H256::zero(), H256::zero(), 0),
+        );
+
+        let state_manager: StateManager = Default::default();
```
Comment (shogochiai) [inline core/src/state/state_manager.rs]: Why is this very abstract trait is generating a concrete state manager?

Comment (syuhei176) [reply core/src/state/state_manager.rs]: using Default trait
https://doc.rust-lang.org/std/default/trait.Default.html

Comment (shogochiai) [reply core/src/state/state_manager.rs]: I understood. This is a default value generator for a complex struct.

File: core/src/state/state_manager.rs (hunk 2 above)
Comment (shogochiai) [reply core/src/state/state_manager.rs]: You mean, if the `StateObject's predicateAddress` is same for the all `outputs(=StateUpdates)` of a Tx, and the predicate says "Okay, this StateObject is mergeable", then we can merge that StateUpdates?

I'm still not sure the apparent difference between StateObject and StateUpdate.

Comment (syuhei176) [reply core/src/state/state_manager.rs]: Basically, we merge StateUpdate, not StateObject. the spec is unstable now but basically, I believe so.
And I suppose whether we can merge state object or not depends on predicate(I guess related to predicate standard).

**Pull request merged**